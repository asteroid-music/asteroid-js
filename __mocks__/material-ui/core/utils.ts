function createSvgIcon(path,displayName) {
    return "Memo("+displayName+"Icon)";
}

export { createSvgIcon };
