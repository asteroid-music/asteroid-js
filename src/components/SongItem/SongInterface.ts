/**
 * Interface for information on a song.
 * Note that songs pulled from get requests to "/queue" also have a
 * 'votes' property whereas other songs may not.
 * Details for each individual property is given in comments below.
 */
interface SongInterface {
    /** Name of the song's artist */
    artist: string,

    /** Name of the song */
    song: string,

    /** Duration of the song, in seconds (float if necessary) */
    duration: number,

    /** Name of the album that the song comes from */
    album: string,

    /** Unique ID that the song is stored under in the backend database */
    _id: string,

    /**
     * Number of votes cast for the particular song.
     * Optional, as only given for songs in the current play queue.
     */
    votes?: number
}

export default SongInterface;
