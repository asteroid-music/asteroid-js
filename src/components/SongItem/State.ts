/**
 * State interface for the <SongItem> component
 * Details for each individual property is given in comments below.
 */
interface SongItemState {
    /** Whether to display song information on separate rows **/
    unfolded: boolean,

    /**
     * The change in vote for this song due to the user since this component
     * was created. I.e the value 0 if the user hasn't voted on this song, 1 if
     * they have upvoted this song, and -1 if they have downvoted this song.
     */
    voteState: number
}

export default SongItemState;
