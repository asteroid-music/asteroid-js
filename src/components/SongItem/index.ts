//Main component
export { AsteroidSongItem, SongInterface } from './SongItem';


