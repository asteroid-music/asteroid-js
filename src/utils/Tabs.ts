/**
 * Interface describing an object with the same elements as SubTabObject
 */
interface SubTabObjectLike {
    /** The name of the tab */
    name: string;

    /** A string array containing names of subtabs */
    component: React.ComponentType<any>;
}

/**
 * Interface describing an object with the same elements as TabObject
 */
interface TabObjectLike {
    /** The name of the tab */
    name: string;

    /** A subtab array containing names of subtabs */
    subtabs: SubTabObjectLike[];
}

class SubTabObject {
    /** The name of the subtab */
    name: string;

    /** The component of the subtab */
    component: React.ComponentType<any>;

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: SubTabObjectLike) {
        this.name = rawJson.name;
        this.component = rawJson.component;
    }

}

/**
 * Class describing named tab, along with strings representing all the subtabs
 * it contains
 */
class TabObject {
    /** The name of the tab */
    name: string;

    /** A string array containing names of subtabs */
    subtabs: SubTabObject[];

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: TabObjectLike) {
        this.name = rawJson.name;
        this.subtabs = rawJson.subtabs.map((item: SubTabObjectLike) => {
            return new SubTabObject(item);
        });
    }

    /**
     * Gets a list of names of subtabs in the tab
     *
     * @returns {string[]} tablist: the name of each subtab
     */
    nameList() {
       return this.subtabs.map((subtab: SubTabObject) => {
            return subtab.name;
       });
    }

    /**
     * Checks if the subtab is present in the tab
     *
     * @param {string} subtab: the name of the subtab to check for
     *
     * @returns {boolean} isin: true if subtab is in the tab; false otherwise
     */
    includes(subtab:string) {
        return this.nameList().includes(subtab);
    }

    /**
     * Returns the subtab specified by the given subtab name,
     *      or 'null' if none exists
     *
     * @param {string} subtab: the name of the subtab to get
     *
     * @returns {SubTabObject | null} subtab: the specified subtab if exists;
     *      'null' otherwise
     */
    get(subtab:string) {
        let objIndex: number = this.nameList().indexOf(subtab);
        if (objIndex === -1) {
            return null;
        } else {
            return this.subtabs[objIndex];
        }
    }
}

/**
 * Class describing an array of named tabs
 */
class TabArray {
    /** Encapsulated array */
    tabs: TabObject[];

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: TabObjectLike[]) {
        this.tabs = rawJson.map((item: TabObjectLike) => {
            return new TabObject(item);
        });
    }

    /**
     * Gets a list of names of tabs in the array
     *
     * @returns {string[]} tablist: the name of each tab
     */
    nameList() {
       return this.tabs.map((tab: TabObject) => {
            return tab.name;
       });
    }

    /**
     * Checks if the tab is present in the array
     *
     * @param {string} tab: the name of the tab to check for
     *
     * @returns {boolean} isin: true if tab is in the tab array;
     *      false otherwise
     */
    includes(tab:string) {
        return this.nameList().includes(tab);
    }

    /**
     * Returns the tab specified by the given tabname, or 'null' if none exists
     *
     * @param {string} tab: the name of the tab to get
     *
     * @returns {TabObject | null} tab: the specified tab if exists;
     *      'null' otherwise
     */
    get(tab:string) {
        let objIndex: number = this.nameList().indexOf(tab);
        if (objIndex === -1) {
            return null;
        } else {
            return this.tabs[objIndex];
        }
    }
}

export { TabObjectLike, TabObject, TabArray }
