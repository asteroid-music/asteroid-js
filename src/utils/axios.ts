import axios from 'axios';

//console.log(asteroidConfig);

declare var __config__: {
    "axios": {
        "baseURL": string,
        "timeout": number
    }
};

let instance = axios.create(__config__.axios || {});

export default instance;
