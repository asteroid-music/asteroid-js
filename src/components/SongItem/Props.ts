//Imports from local directory
import SongInterface from './SongInterface';

/**
 * Props interface for the <SongItem> component
 * Details for each individual property is given in comments below.
 */
interface SongItemProps {
    /** The song to display information for - see SongInterface **/
    song: SongInterface,

    /** Whether to show upvote and downvote buttons for the song **/
    voteButtons?: boolean,

    /** Whether to initially render the SongItem unfolded **/
    unfolded?: boolean
}

export default SongItemProps;
