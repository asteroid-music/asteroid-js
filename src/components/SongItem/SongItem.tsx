//Imports from external 'react' module
import React from 'react';
import type { ElementType } from 'react';

//Imports from external '@material-ui/core' module
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Typography from '@material-ui/core/Typography';

//Imports from external '@material-ui/core/colors' module
import green from '@material-ui/core/colors/green';
import red from '@material-ui/core/colors/red';

//Imports from external '@material-ui/icons' module
import ArrowUpwardIcon from '@material-ui/icons/ArrowUpward';
import ArrowDownwardIcon from '@material-ui/icons/ArrowDownward';
import UnfoldMore from '@material-ui/icons/UnfoldMore';
import UnfoldLess from '@material-ui/icons/UnfoldLess';
import IconButton from '@material-ui/core/IconButton';

//Import from external 'axios' module
import axios from 'utils/axios';

//Imports from local directory
import SongInterface from './SongInterface';
import SongItemProps from './Props';
import SongItemState from './State';

/**
 * <SongItem> component allowing interaction
 * with a single song item.
 */
class SongItem extends React.Component<SongItemProps,SongItemState> {
    constructor(props) {
        let propsAreValid: boolean = true;
        //Assert that the only props are 'song', 'voteButtons' and 'unfolded'
        Object.keys(props).forEach(key=>{
            if (key !== "song" && key !== "voteButtons" && key !== "unfolded") {
                propsAreValid = false;
            }
        });
        //Assert that the 'song' prop exists and includes the correct entries
        if (props.song) {
            //Assert that the only elements are 'artist', 'song', 'duration', 'album', '_id' and 'votes'
            Object.keys(props.song).forEach(key=>{
                if (!["artist","song","duration","album","_id","votes"].includes(key)) {
                    propsAreValid = false;
                }
            });
            //Assert that types are matched
            propsAreValid = propsAreValid && (typeof props.song.artist === "string" || props.song.artist instanceof String);
            propsAreValid = propsAreValid && (typeof props.song.song === "string" || props.song.song instanceof String);
            propsAreValid = propsAreValid && (typeof props.song.album === "string" || props.song.album instanceof String);
            propsAreValid = propsAreValid && (typeof props.song._id === "string" || props.song._id instanceof String);
            propsAreValid = propsAreValid && (typeof props.song.duration === "number" || props.song.duration instanceof Number);
            propsAreValid = propsAreValid && (!props.song.votes || (typeof props.song.votes === "number" || props.song.votes instanceof Number));
        } else {propsAreValid = false;}
        if(!propsAreValid) {
            throw Error("Unable to render the <SongItem> component: invalid props. This is likely an issue with the server and not your browser.")
        }
        super(props);
        this.state = {
            unfolded: this.props.unfolded,
            voteState: 0
        };
    }

    unfoldedSecondaryNode() {
        let song = this.props.song;
        let dur = this._asTime(song.duration);
        return <div>
            <Typography>Artist: {song.artist}</Typography>
            <Typography>Album: {song.album}</Typography>
            <Typography>Duration: {dur}</Typography>
        </div>
    }

    foldedSecondaryText() {
        let song = this.props.song;
        return song.artist + " - " + song.album + " - " + this._asTime(song.duration);
    }

    _vote(value: number) {
        const voteState = this.state.voteState;
        const postUrl="/queue?song_id="+this.props.song._id+"&vote="+value;
        axios.post<{message: string}>(postUrl,{}).catch(
            (error) => {
                this.setState({voteState:voteState});
            }
        );
    }

    _asTime(seconds: number) {
        let secs: number = seconds % 60;
        let mins: number = (seconds-secs)/60;
        return mins.toString()+":"+secs.toString();
    }

    upvoteCallback() {
        let voteState = this.state.voteState;
        if (voteState == 1) {
            this._vote(-1);
            this.setState({voteState:0});
        } else {
            this._vote(1-voteState);
            this.setState({voteState:1});
        }
    }

    downvoteCallback() {
        let voteState = this.state.voteState;
        if (voteState == -1) {
            this._vote(1);
            this.setState({voteState:0});
        } else {
            this._vote(voteState-1);
            this.setState({voteState:-1});
        }
    }

    render () {
        let song = this.props.song;

        let unfolded = this.state.unfolded;
        let voteState = this.state.voteState;

        let textSecondary = unfolded
            ? this.unfoldedSecondaryNode()
            : this.foldedSecondaryText();

        let foldButton = unfolded
            ? <IconButton onClick={() => {this.setState({unfolded:false});}}><UnfoldLess /></IconButton>
            : <IconButton onClick={() => {this.setState({unfolded:true});}}><UnfoldMore /></IconButton>

        let otherProps = unfolded
            ? {
                secondaryTypographyProps: {component:"div" as ElementType}
            }
            : {
                primaryTypographyProps: {noWrap:true},
                secondaryTypographyProps: {noWrap:true}
            };

        let upButtonStyle = (voteState === 1)
            ? {backgroundColor: green[100]}
            : {};

        let downButtonStyle = (voteState === -1)
            ? {backgroundColor: red[100]}
            : {};

        let upButton = this.props.voteButtons && <IconButton
            style={upButtonStyle}
            onClick={() => {this.upvoteCallback();}}
        >
            <ArrowUpwardIcon
                style={{ color: green[500] }}
            />
        </IconButton>

        let downButton = this.props.voteButtons && <IconButton
            style={downButtonStyle}
            onClick={() => {this.downvoteCallback();}}
        >
            <ArrowDownwardIcon
                style={{ color: red[500] }}
            />
        </IconButton>

        let votesInfo = song.votes && <Typography>{song.votes+voteState}</Typography>

        return <ListItem>
            <ListItemText
                primary={song.song}
                secondary={textSecondary}
                {...otherProps}
            />
            {upButton}
            {votesInfo}
            {downButton}
            {foldButton}
        </ListItem>
    }
}

export default SongItem;
