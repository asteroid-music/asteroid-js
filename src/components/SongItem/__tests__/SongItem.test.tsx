import React from 'react';
import renderer from 'react-test-renderer';
import mockAxios from 'jest-mock-axios';
import SongItem, {SongInterface} from 'components/SongItem';
import { serialize as serialise } from 'jest-snapshot/build/utils';
//I know, UK vs US spelling...

/**
 * SongItem.tsx imports:
 *  - React
 *  - Components from MUI (pre-mocked)
 *  - Colours and icons from MUI (pre-mocked)
 *  - axios
 *  - SongInterface
 * 
 * The following should be mocked:
 *  - axios
 *  - SongInterface
 * 
 * As this component makes a POST request to the
 *  /queue endpoint, it should be tested for behaviour
 *  with all possible responses:
 * 
 *  - 201 Successful response
 *  - 400 Bad Request (TODO)
 *  - 422 Validation Error (TODO)
 * 
 * As this component contains the following interactions:
 *  - Fold/unfold button
 *  - Upvote button
 *  - Downvote button
 * it should test the interaction with each
 */

afterEach(()=>{mockAxios.reset();});

//Test positive SongInterface without votes
const testSongVoteless: SongInterface = {
    artist: "This is a test artist name",
    song: "This is a test song name",
    duration: 556,
    album: "This is a test album name",
    _id: "aaaaaaa"
}

//Test negative SongInterface with votes
const testSongVoteful: SongInterface = {
    votes: 3,
    ...testSongVoteless
}

//Test correct props but invalid data
const invalidSongData = {
    invalidSongData: true
} as unknown as SongInterface; //Bypasses TS type check to test incorrectly typed data

//Test incorrect props
const invalidPropsData = {
    crunch: "hello"
} as unknown as {song: SongInterface}; //Bypasses TS type check to test incorrectly typed data

describe('<SongItem />',()=>{
    //Basic rendering with minimal props
    describe('basic rendering',()=>{
        it("renders correctly with a voteless song",()=>{
            expect(renderer
                .create(<SongItem song={testSongVoteless} />)
                .toJSON()
            ).toMatchSnapshot();
        });
        it("renders correctly with a voteful song",()=>{
            expect(renderer
                .create(<SongItem song={testSongVoteful} />)
                .toJSON()
            ).toMatchSnapshot();
        });
        it("throws an error with missing/invalid props",()=>{
            console.error = jest.fn();
            expect(()=>{
                renderer
                .create(<SongItem {...invalidPropsData} />);
            }).toThrowErrorMatchingSnapshot();
        });
        it("throws an error if the 'song' prop is not a valid SongItem",()=>{
            console.error = jest.fn();
            expect(()=>{
                renderer
                .create(<SongItem song={invalidSongData} />);
            }).toThrowErrorMatchingSnapshot();
        });
    });

    //Basic rendering with voteButtons props
    describe('prop: voteButtons',()=>{
        it("renders correctly with a voteless song and vote buttons",()=>{
            expect(renderer
                .create(<SongItem song={testSongVoteless} voteButtons />)
                .toJSON()
            ).toMatchSnapshot();
        });
        it("renders correctly with a voteful song and vote buttons",()=>{
            expect(renderer
                .create(<SongItem song={testSongVoteful} voteButtons />)
                .toJSON()
            ).toMatchSnapshot();
        });
    });

    //Testing unfolded
    describe('prop: unfolded',()=>{
        it("renders correctly when unfolded by default",()=>{
            expect(renderer
                .create(<SongItem song={testSongVoteless} unfolded />)
                .toJSON()
            ).toMatchSnapshot();
        });

        it("is identical"
            + " to when there is no unfolded prop"
            + " and then the unfold button is pressed",
            ()=>{
            //Render folded and unfolded with test renderer
            let unfoldedTree = renderer.create(<SongItem song={testSongVoteless} unfolded/>);
            let foldedTree = renderer.create(<SongItem song={testSongVoteless} />);
            //Naively interact with first button on folded element, unfolding
            /** Assumptions:
             * 1. Clicking successfully calls onClick event (tested within mui)
             * 2. There is only one button - the unfold button (tested by prior tests,
             *      presuming valid snapshots to check against
             */
            foldedTree.root.findByType('button').props.onClick();
            //Generate serialised form
            let unfoldedSerialised = serialise(unfoldedTree.toJSON());
            let foldedSerialised = serialise(foldedTree.toJSON());
            //Expect to equal unfolded
            expect(foldedSerialised).toEqual(unfoldedSerialised);
        });

        it("folds down"
            + " when the fold button is pressed"
            + " to match when there is no unfolded prop",
            ()=>{
            //Render folded and unfolded with test renderer
            let unfoldedTree = renderer.create(<SongItem song={testSongVoteless} unfolded/>);
            let foldedTree = renderer.create(<SongItem song={testSongVoteless} />);
            //Naively interact with first button on unfolded element, folding
            /** Assumptions:
             * 1. Clicking successfully calls onClick event (tested within mui)
             * 2. There is only one button - the fold button (tested by prior tests,
             *      presuming valid snapshots to check against
             */
            unfoldedTree.root.findByType('button').props.onClick();
            //Generate serialised form
            let unfoldedSerialised = serialise(unfoldedTree.toJSON());
            let foldedSerialised = serialise(foldedTree.toJSON());
            //Expect to equal unfolded
            expect(unfoldedSerialised).toEqual(foldedSerialised);
        });

    });

    //Interaction with vote buttons
    describe('interaction: voting',()=>{
        it('increases the number of votes by one'
            + ' when the upvote button is hit'
            + ' and POST request is successful',
            ()=>{
            //Render with test renderer
            let tree = renderer.create(<SongItem song={testSongVoteful} voteButtons />);
            //Naively interact with first button
            /** Assumptions:
             *  Button with index '0' is upvote button (tested by prior tests,
             *      presuming valid snapshots to check against)
             */
            tree.root.findAllByType('button')[0].props.onClick();
            //Check that a post request has been made
            expect(mockAxios.post).toHaveBeenCalledWith("/queue?song_id=aaaaaaa&vote=1",{});
            //Simulate a server response
            mockAxios.mockResponse({data:{message:"ok"}});
            //Expect to match snapshot
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('decreases the number of votes by one'
            + ' when the downvote button is hit'
            + ' and POST request is successful',
            ()=>{
            //Render with test renderer
            let tree = renderer.create(<SongItem song={testSongVoteful} voteButtons />);
            //Naively interact with second button
            /** Assumptions:
             * Button with index '1' is downote button (tested by prior tests,
             *      presuming valid snapshots to check against
             */
            tree.root.findAllByType('button')[1].props.onClick();
            //Check that a post request has been made
            expect(mockAxios.post).toHaveBeenCalledWith("/queue?song_id=aaaaaaa&vote=-1",{});
            //Simulate a server response
            mockAxios.mockResponse({data:{message:"ok"}});
            //Expect to match snapshot
            expect(tree.toJSON()).toMatchSnapshot();
        });
    });
});