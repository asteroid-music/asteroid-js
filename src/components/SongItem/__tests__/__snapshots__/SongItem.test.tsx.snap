// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SongItem /> basic rendering renders correctly with a voteful song 1`] = `
<Material UI ListItem>
  <Material UI ListItemText
    primary="This is a test song name"
    primaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
    secondary="This is a test artist name - This is a test album name - 9:16"
    secondaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
  />
  <Material UI Typography>
    3
  </Material UI Typography>
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(UnfoldMoreIcon) /&gt;,
  "onClick": [Function],
}
  </button>
</Material UI ListItem>
`;

exports[`<SongItem /> basic rendering renders correctly with a voteless song 1`] = `
<Material UI ListItem>
  <Material UI ListItemText
    primary="This is a test song name"
    primaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
    secondary="This is a test artist name - This is a test album name - 9:16"
    secondaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
  />
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(UnfoldMoreIcon) /&gt;,
  "onClick": [Function],
}
  </button>
</Material UI ListItem>
`;

exports[`<SongItem /> basic rendering throws an error if the 'song' prop is not a valid SongItem 1`] = `"Unable to render the <SongItem> component: invalid props. This is likely an issue with the server and not your browser."`;

exports[`<SongItem /> basic rendering throws an error with missing/invalid props 1`] = `"Unable to render the <SongItem> component: invalid props. This is likely an issue with the server and not your browser."`;

exports[`<SongItem /> interaction: voting decreases the number of votes by one when the downvote button is hit and POST request is successful 1`] = `
<Material UI ListItem>
  <Material UI ListItemText
    primary="This is a test song name"
    primaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
    secondary="This is a test artist name - This is a test album name - 9:16"
    secondaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
  />
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(ArrowUpwardIcon)
    style={
      Object {
        "color": "#4caf50",
      }
    }
  /&gt;,
  "onClick": [Function],
  "style": Object {},
}
  </button>
  <Material UI Typography>
    2
  </Material UI Typography>
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(ArrowDownwardIcon)
    style={
      Object {
        "color": "#f44336",
      }
    }
  /&gt;,
  "onClick": [Function],
  "style": Object {
    "backgroundColor": "#ffcdd2",
  },
}
  </button>
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(UnfoldMoreIcon) /&gt;,
  "onClick": [Function],
}
  </button>
</Material UI ListItem>
`;

exports[`<SongItem /> interaction: voting increases the number of votes by one when the upvote button is hit and POST request is successful 1`] = `
<Material UI ListItem>
  <Material UI ListItemText
    primary="This is a test song name"
    primaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
    secondary="This is a test artist name - This is a test album name - 9:16"
    secondaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
  />
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(ArrowUpwardIcon)
    style={
      Object {
        "color": "#4caf50",
      }
    }
  /&gt;,
  "onClick": [Function],
  "style": Object {
    "backgroundColor": "#c8e6c9",
  },
}
  </button>
  <Material UI Typography>
    4
  </Material UI Typography>
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(ArrowDownwardIcon)
    style={
      Object {
        "color": "#f44336",
      }
    }
  /&gt;,
  "onClick": [Function],
  "style": Object {},
}
  </button>
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(UnfoldMoreIcon) /&gt;,
  "onClick": [Function],
}
  </button>
</Material UI ListItem>
`;

exports[`<SongItem /> prop: unfolded renders correctly when unfolded by default 1`] = `
<Material UI ListItem>
  <Material UI ListItemText
    primary="This is a test song name"
    secondary={
      <div>
        <Material UI Typography>
          Artist: 
          This is a test artist name
        </Material UI Typography>
        <Material UI Typography>
          Album: 
          This is a test album name
        </Material UI Typography>
        <Material UI Typography>
          Duration: 
          9:16
        </Material UI Typography>
      </div>
    }
    secondaryTypographyProps={
      Object {
        "component": "div",
      }
    }
  />
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(UnfoldLessIcon) /&gt;,
  "onClick": [Function],
}
  </button>
</Material UI ListItem>
`;

exports[`<SongItem /> prop: voteButtons renders correctly with a voteful song and vote buttons 1`] = `
<Material UI ListItem>
  <Material UI ListItemText
    primary="This is a test song name"
    primaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
    secondary="This is a test artist name - This is a test album name - 9:16"
    secondaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
  />
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(ArrowUpwardIcon)
    style={
      Object {
        "color": "#4caf50",
      }
    }
  /&gt;,
  "onClick": [Function],
  "style": Object {},
}
  </button>
  <Material UI Typography>
    3
  </Material UI Typography>
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(ArrowDownwardIcon)
    style={
      Object {
        "color": "#f44336",
      }
    }
  /&gt;,
  "onClick": [Function],
  "style": Object {},
}
  </button>
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(UnfoldMoreIcon) /&gt;,
  "onClick": [Function],
}
  </button>
</Material UI ListItem>
`;

exports[`<SongItem /> prop: voteButtons renders correctly with a voteless song and vote buttons 1`] = `
<Material UI ListItem>
  <Material UI ListItemText
    primary="This is a test song name"
    primaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
    secondary="This is a test artist name - This is a test album name - 9:16"
    secondaryTypographyProps={
      Object {
        "noWrap": true,
      }
    }
  />
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(ArrowUpwardIcon)
    style={
      Object {
        "color": "#4caf50",
      }
    }
  /&gt;,
  "onClick": [Function],
  "style": Object {},
}
  </button>
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(ArrowDownwardIcon)
    style={
      Object {
        "color": "#f44336",
      }
    }
  /&gt;,
  "onClick": [Function],
  "style": Object {},
}
  </button>
  <button
    onClick={[Function]}
  >
    Material UI IconButton with props:
    Object {
  "children": &lt;Memo(UnfoldMoreIcon) /&gt;,
  "onClick": [Function],
}
  </button>
</Material UI ListItem>
`;
